From cb5cf0215c42375aa4c3e6377a6a006e2969846d Mon Sep 17 00:00:00 2001
From: Nicolas Boulenguez <nicolas@debian.org>
Date: Wed, 10 Jan 2024 20:24:18 +0100
Subject: [PATCH 05/10] scripts/architecture.mk: reduce the number of
 subprocesses

Ensure that dpkg-architecture is only called when a
DEB_{HOST,ARCH_TARGET}_* variable is is expanded.

If so, call it once and cache all results for future expansions.
---
 scripts/mk/architecture.mk | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/scripts/mk/architecture.mk b/scripts/mk/architecture.mk
index 2ffcee287..174c5b638 100644
--- a/scripts/mk/architecture.mk
+++ b/scripts/mk/architecture.mk
@@ -5,12 +5,20 @@
 ifndef dpkg_architecture.mk_included
 dpkg_architecture.mk_included :=
 
-dpkg_lazy_eval ?= $$(or $$(value DPKG_CACHE_$(1)),$$(eval DPKG_CACHE_$(1) := $$(shell $(2)))$$(value DPKG_CACHE_$(1)))
-
-dpkg_architecture_setvar = export $(1) ?= $(call dpkg_lazy_eval,$(1),dpkg-architecture -q$(1))
-
+dpkg-architecture_vars := \
 $(foreach machine,BUILD HOST TARGET,\
   $(foreach var,ARCH ARCH_ABI ARCH_LIBC ARCH_OS ARCH_CPU ARCH_BITS ARCH_ENDIAN GNU_CPU GNU_SYSTEM GNU_TYPE MULTIARCH,\
-    $(eval $(call dpkg_architecture_setvar,DEB_$(machine)_$(var)))))
+    DEB_$(machine)_$(var)))
+
+# Spare a subprocess in the frequent case where dpkg-buildpackage
+# has already exported all variables.
+ifneq (,$(filter undefined,$(foreach v,$(dpkg-architecture_vars),$(origin $(v)))))
+
+  # ?= preserves overriddes from debian/rules or its command line.
+  $(foreach v,$(subst =,?=,$(shell dpkg-architecture)),$(eval $(v)))
+
+endif
+
+export $(dpkg-architecture_vars)
 
 endif
-- 
2.39.2

